server/roomlogs.ts

```diff
@@ -222,42 +239,73 @@ export class Roomlog {
 		this.broadcastBuffer.push(fullMessage);
 	}
 	parseChatLine(line: string) {
-		const messageStart = !this.noLogTimes ? '|c:|' : '|c|';
-		const section = !this.noLogTimes ? 4 : 3; // ['', 'c' timestamp?, author, message]
-		if (line.startsWith(messageStart)) {
-			const parts = Utils.splitFirst(line, '|', section);
-			return {user: parts[section - 1], message: parts[section]};
+		const prefixes = [['|c:|', 4], ['|c|', 3]] as [string, number][];
+		for (const [messageStart, section] of prefixes) {
+			// const messageStart = !this.noLogTimes ? '|c:|' : '|c|';
+			// const section = !this.noLogTimes ? 4 : 3; // ['', 'c' timestamp?, author, message]
+			if (line.startsWith(messageStart)) {
+				const parts = Utils.splitFirst(line, '|', section);
+				return {user: parts[section - 1], message: parts[section]};
+			}
 		}
 	}
 	roomlog(message: string, date = new Date()) {
-		if (!this.roomlogStream) return;
-		const timestamp = Chat.toTimestamp(date).split(' ')[1] + ' ';
 		message = message.replace(/<img[^>]* src="data:image\/png;base64,[^">]+"[^>]*>/g, '');
-		void this.roomlogStream.write(timestamp + message + '\n');
+		if (roomlogTable && !(!Config.logchat || this.roomid.startsWith('battle-'))) {
+			const chatData = this.parseChatLine(message);
+			const type = message.split('|')[1] || "";
+			const insertQuery = SQL`INSERT INTO roomlogs (type, roomid, userid, time, log)`;
+			insertQuery.append(SQL` VALUES (${type}, ${this.roomid}, ${toID(chatData?.user) || null}, `);
+			insertQuery.append(SQL`${date.getTime()}, ${message})`);
+			void this.insertLog(insertQuery);
+		} else if (this.roomlogStream) {
+			const timestamp = Chat.toTimestamp(date).split(' ')[1] + ' ';
+			void this.roomlogStream.write(timestamp + message + '\n');
+		}
+	}
+	private async insertLog(query: SQLStatement): Promise<void> {
+		if (!roomlogTable) return;
+		try {
+			await roomlogTable.query(query);
```

Comment:
```suggestion
		try {
			await roomlogTable?.query(query);
```

Reply:
I left this check in since it makes the non-null assertions for `replaysDB` safe to use down a few lines.